### PR TITLE
Introduce metrics column visibility

### PR DESCRIPTION
**Pull request by KKoukiou**
# Metrics page: control visibility of the resource usage graphs

Screen Shot 2023-02-07 at 14 13 59

### CODE REVIEW

**Comment by garrett** [comment]
Since we have multiple items that can be toggled, I think we might want to use checkbox select instead:

- Design: https://www.patternfly.org/v4/demos/filters/design-guidelines#checkbox-select
- React: https://www.patternfly.org/v4/components/select#checkbox-input-with-counts
- React next: https://www.patternfly.org/v4/components/select/react-next#checkbox
- HTML: https://www.patternfly.org/v4/components/select/html/#checkbox

React has versions with a count badge and also without. Should we have a count badge?

File: pkg/metrics/metrics.jsx [hunk 1]
```diff
@@ -1667,33 +1676,68 @@ class MetricsHistory extends React.Component {
...
+            <Select
+                toggleAriaLabel={_("Graph visibility options menu")}
+                className="metrics-label"
+                variant={SelectVariant.checkbox}
+                isCheckboxSelectionBadgeHidden
+                isOpen={!!this.state.isOpenColumnVisibility}
+                onSelect={(_, selection) => {
+                    const s = this.columns.find(itm => itm[1] == selection);
+                    this.setState({ selectedVisibility: { ...this.state.selectedVisibility, [s[0]]: !this.state.selectedVisibility[s[0]] } });
+                }}
```
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:1702]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18299-20230208-135320-0fc62071-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx [hunk 2]
```diff
@@ -1667,33 +1676,68 @@
...
+                            {graphsVisibilitySelector}
+                        </Flex>
                         <div className="metrics-graphs">
-                            <Label label={_("CPU")} items={[_("Usage"), _("Load")]} />
-                            <Label label={_("Memory")} items={[_("Usage"), ...swapTotal ? [_("Swap")] : []]} />
-                            <Label label={_("Disk I/O")} items={[_("Usage")]} />
-                            <Label label={_("Network")} items={[_("Usage")]} />
+                            {this.state.selectedVisibility.cpu && <Label label={_("CPU")} items={[_("Usage"), _("Load")]} />}
+                            {this.state.selectedVisibility.memory && <Label label={_("Memory")} items={[_("Usage"), ...swapTotal ? [_("Swap")] : []]} />}
+                            {this.state.selectedVisibility.disk && <Label label={_("Disk I/O")} items={[_("Usage")]} />}
```
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:1731]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18299-20230208-135320-0fc62071-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx [hunk 3]
```diff
@@ -1055,22 +1055,29 @@ class MetricsHour extends React.Component {
...
         return true;
     }
 
     // data: type → SAMPLES_PER_H objects from startTime
-    updateGraphs(data, startTime) {
+    updateGraphs(data, startTime, selectedVisibility) {
+        const filteredData = data.map(sample => Object.keys(sample)
+                .filter(key => {
+                    return selectedVisibility[key.split("_")[1]];
+                })
```
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:1077]: This is better written as a pure arrow, i.e. `key => selectedVisibility[key.split("_")[1]]`

File: pkg/metrics/metrics.jsx [hunk 4]
```diff
@@ -1667,33 +1676,68 @@ class MetricsHistory extends React.Component {
...
+                                aria-label={_("Jump to")}
+                                onToggle={this.handleToggle}
+                                onSelect={this.handleSelect}
+                                isOpen={this.state.isDatepickerOpened}
+                                selections={this.state.selectedDate}
+                                toggleId="date-picker-select-toggle"
+                            >
+                                {options}
+                            </Select>
+                            {graphsVisibilitySelector}
```
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:1726]: Feels a bit inconsistent to define the date picker inline here, and the graph visibility selector separately, but no biggie.

**Review by martinpitt — APPROVED** [review]
Nice work! Code-wise this looks fine (feel free to ignore the two comments, they are not at all important). If @garrett is happy with the design, so am I.

File: pkg/metrics/metrics.jsx (hunk 1 above)
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:1702]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18299-20230208-164058-692abce6-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:1731]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18299-20230208-164058-692abce6-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx (hunk 1 above)
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:1702]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18299-20230208-180314-3dd085b5-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:1731]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18299-20230208-180314-3dd085b5-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx (hunk 4 above)
Comment (KKoukiou) [reply pkg/metrics/metrics.jsx]: Moved.

**Comment by garrett** [comment]
So I'm not sure if this is fine enough or not. It's probably fine, as one of the points is a rendering oddity (alignment) and then there are corner cases when everything is turned off.

But this PR does allow people to toggle specific parts off, which is the intention. And turning them back on is also intentional.

The only other thing is that we could have a "Enable all" or "Reset to default" action. But there aren't so many options as to make this fully needed.

---

Here's everything on:

Screenshot 2023-02-09 at 12-07-33 Overview - admin@fedora-37-127-0-0-2-2201

When there's just one item, the header seems aligned oddly:

Screenshot 2023-02-09 at 12-07-49 Overview - admin@fedora-37-127-0-0-2-2201

Spikes are connected. I had a network spike. Turning off the graph hides the spike:

Screenshot 2023-02-09 at 12-08-18 Overview - admin@fedora-37-127-0-0-2-2201

When nothing is selected, you see nothing. Not even the logs.

Screenshot 2023-02-09 at 12-08-00 Overview - admin@fedora-37-127-0-0-2-2201

**Comment by martinpitt** [comment]
Hiding events when turning off their corresponding graph seems intended to me. However, turning everything off and ending up with a weirdly broken page seems worth fixing? I.e. can we just disallow turning off everything? Perhaps by disabling the click on the last enabled item? Or we don't care enough?

(Code-wise, my +1 stands; thanks for the little cleanups!)

**Comment by KKoukiou** [comment]
> Hiding events when turning off their corresponding graph seems intended to me. However, turning everything off and ending up with a weirdly broken page seems worth fixing? I.e. can we just disallow turning off everything? Perhaps by disabling the click on the last enabled item? Or we don't care enough?
> 
> (Code-wise, my +1 stands; thanks for the little cleanups!)

I was thinking about disabling the last enabled item as well when developing this. I will merge this as is, and discuss about it with @garrett garrett again on monday.

**Pull request merged**